walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,107 @@
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+
 
 public final class FindMeetingQuery {
+
+  /*
+   Criterion functional interface is used to define who's (mandatory or all attendees')
+   occupied hours should be taken into account in the findAvailableTimeRanges function.
+   */
+  @FunctionalInterface
+  private interface Criterion {
+    boolean satisfies(int x);
+  }
+  private final Criterion COMMON_CRITERION = x -> x == 0;
+  private final Criterion MANDATORY_CRITERION = x -> x <= 1;
+
+  private static final int MINUTES = 60 * 24;
+
+  private int getStart(Event event) {
+    return event.getWhen().start();
+  }
+
+  private int getEnd(Event event) {
+    return event.getWhen().end();
+  }
+
+  private boolean haveCommonAttendees(Collection<String> a, Collection<String> b) {
+    return new HashSet<>(a).removeAll(b);
+  }
+
+  /*
+  The algorithm skips all minutes that don't meet the criterion,
+  memorises the index of the first fitting minute in a row and goes until
+  it finds the new unfitting minute. Adds the range to the list if
+  it's not less than the target duration.
+   */
+  /**
+   * @param occupied shows what minutes are available for booking
+   *                 for mandatory (0|1) or all(0) attendees
+   * @param criterion defines whose (mandatory or all attendees')
+   *                  occupied hours should be taken in account
+   * @param duration the duration of the new meeting
+   * @return the list of time ranges available to schedule the meeting
+   * */
+  private List<TimeRange> findAvailableTimeRanges(int [] occupied, final Criterion criterion, int duration) {
+    List<TimeRange> timeRanges = new ArrayList<>();
+    int index = 0;
+    while (index < MINUTES) {
+      int start = index;
+      while (index < MINUTES && criterion.satisfies(occupied[index])) {
+        index++;
+      }
+      if (index - start >= duration) {
+        timeRanges.add(TimeRange.fromStartDuration(start, index - start));
+      }
+      while (index < MINUTES && !criterion.satisfies(occupied[index])) {
+        index++;
+      }
+    }
+    return timeRanges;
+  }
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    if (events == null) {
+      throw new IllegalArgumentException("Events list should be a non-null value");
+    }
+    if (request == null) {
+      throw new IllegalArgumentException("Request should be a non-null value");
+    }
+    Collection<String> mandatoryAttendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    int duration = (int) request.getDuration();
+
+    /*
+    occupied array indicates if the i-th hour is occupied
+    for mandatory (2) or optional (1) attendee
+    or not occupied (0)
+     */
+    int [] occupied = new int[MINUTES + 1];
```

Comment:
nit: remove the space between `int` and `[]` per [style guidelines](https://google.github.io/styleguide/javaguide.html#s4.6.2-horizontal-whitespace)